File: packages/soul-common/src/soul_common/catalog/catalog_service.py
"""
CatalogService - 폴더/세션 카탈로그 관리 서비스 계층

REST API와 MCP 도구가 공용으로 사용하는 비즈니스 로직 계층.
DB 호출 + 브로드캐스트를 캡슐화하여 구현 중복을 제거한다.
"""

import uuid
from typing import Any, Optional, Protocol, runtime_checkable

from soul_common.db.session_db import PostgresSessionDB

_UNSET = object()
BOARD_GRID_SIZE = 20
BOARD_TILE_WIDTH = 160
BOARD_TILE_HEIGHT = 120
BOARD_DEFAULT_COLUMNS = 4


@runtime_checkable
class SessionBroadcasterProtocol(Protocol):
    """세션 목록 변경 브로드캐스터 인터페이스.

    CatalogService가 의존하는 최소 인터페이스.
    soul-server의 SessionBroadcaster, soulstream-server의 SoulstreamBroadcaster 등
    각 서비스가 자체 구현을 제공한다.
    """

    async def broadcast(self, event: dict) -> int: ...

    async def emit_session_deleted(self, agent_session_id: str) -> int: ...


class CatalogService:
    """폴더/세션 카탈로그 관리 서비스"""

    def __init__(
        self,
        session_db: PostgresSessionDB,
        broadcaster: SessionBroadcasterProtocol,
    ):
        self._db = session_db
        self._broadcaster = broadcaster

    async def _broadcast_catalog(self) -> None:
        """카탈로그 변경을 모든 리스너에게 브로드캐스트한다."""
        catalog = await self._db.get_catalog()
        await self._broadcaster.broadcast({
            "type": "catalog_updated",
            "catalog": catalog,
        })

    def _snap_position(self, value: float) -> float:
        return float(round(value / BOARD_GRID_SIZE) * BOARD_GRID_SIZE)

    async def _next_board_position(self, folder_id: str) -> tuple[float, float]:
        await self._db.ensure_board_items()
        items = [
            item for item in await self._db.get_board_items()
            if item.get("folderId") == folder_id
        ]
        occupied = {
            (int(item.get("x", 0)), int(item.get("y", 0)))
            for item in items
        }
        index = 0
        while True:
            x = (index % BOARD_DEFAULT_COLUMNS) * BOARD_TILE_WIDTH
            y = (index // BOARD_DEFAULT_COLUMNS) * BOARD_TILE_HEIGHT
            if (x, y) not in occupied:
                return float(x), float(y)
            index += 1

    # --- 폴더 CRUD ---

    async def list_folders(self) -> list[dict]:
        """전체 폴더 목록을 반환한다."""
        folders = await self._db.get_all_folders()
        return [
            {
                "id": f["id"],
                "name": f["name"],
                "sortOrder": f["sort_order"],
                "parentFolderId": f.get("parent_folder_id"),
                "settings": f.get("settings") or {},
                "createdAt": f.get("created_at"),
            }
            for f in folders
        ]

    async def create_folder(
        self,
        name: str,
        sort_order: int = 0,
        parent_folder_id: Optional[str] = None,
    ) -> dict:
        """폴더를 생성하고 결과를 반환한다."""
        folder_id = str(uuid.uuid4())
        await self._validate_parent_folder(folder_id, parent_folder_id)
        await self._db.create_folder(
            folder_id,
            name,
            sort_order,
            parent_folder_id=parent_folder_id,
        )
        await self._broadcast_catalog()
        return {
            "id": folder_id,
            "name": name,
            "sortOrder": sort_order,
            "parentFolderId": parent_folder_id,
            "settings": {},
        }

    async def rename_folder(self, folder_id: str, name: str) -> None:
        """폴더 이름을 변경한다."""
        await self._db.update_folder(folder_id, name=name)
        await self._broadcast_catalog()

    async def update_folder(
        self,
        folder_id: str,
        name: Optional[str] = None,
        sort_order: Optional[int] = None,
        settings: Optional[dict] = None,
        parent_folder_id: Any = _UNSET,
    ) -> None:
        """폴더의 이름, 정렬 순서, 설정을 변경한다."""
        fields: dict = {}
        if name is not None:
            fields["name"] = name
        if sort_order is not None:
            fields["sort_order"] = sort_order
        if settings is not None:
            fields["settings"] = settings
        if parent_folder_id is not _UNSET:
            await self._validate_parent_folder(folder_id, parent_folder_id)
            fields["parent_folder_id"] = parent_folder_id
        if not fields:
            return
        await self._db.update_folder(folder_id, **fields)
        await self._broadcast_catalog()

    async def list_child_folders(self, folder_id: Optional[str]) -> list[dict]:
        folders = await self.list_folders()
        return [f for f in folders if f.get("parentFolderId") == folder_id]

    async def delete_folder(self, folder_id: str) -> None:
        """폴더를 삭제한다."""
        await self._db.delete_folder(folder_id)
        await self._broadcast_catalog()

    async def reorder_folders(self, items: list[dict]) -> None:
        """여러 폴더의 sort_order와 parent_folder_id를 한 번에 업데이트한다.

        Args:
            items: [{"id": str, "sortOrder": int, "parentFolderId"?: str | None}, ...] 형태의 목록
        """
        await self._validate_folder_parent_updates({
            item["id"]: item["parentFolderId"]
            for item in items
            if "parentFolderId" in item
        })
        for item in items:
            fields: dict[str, Any] = {"sort_order": item["sortOrder"]}
            if "parentFolderId" in item:
                fields["parent_folder_id"] = item["parentFolderId"]
            await self._db.update_folder(item["id"], **fields)
        await self._broadcast_catalog()

    # --- 세션 관리 ---

    async def move_sessions_to_folder(
        self,
        session_ids: list[str],
        folder_id: Optional[str],
    ) -> None:
        """세션들을 지정한 폴더로 이동한다. folder_id가 None이면 미배정."""
        for sid in session_ids:
            await self._db.assign_session_to_folder(sid, folder_id)
        await self._db.ensure_board_items()
        await self._broadcast_catalog()

    async def broadcast_catalog(self) -> None:
        """카탈로그 상태를 브로드캐스트한다. DB 변경 없이 broadcast만 수행."""
        await self._broadcast_catalog()

    async def rename_session(
        self,
        session_id: str,
        display_name: Optional[str],
    ) -> None:
        """세션의 표시 이름을 변경한다."""
        await self._db.rename_session(session_id, display_name)
        await self._broadcast_catalog()

    async def delete_session(self, session_id: str) -> None:
        """세션을 삭제한다."""
        await self._db.delete_session(session_id)
        # catalog_updated로 대시보드 폴더 뷰 갱신
        await self._broadcast_catalog()
        # session_deleted로 세션 목록 뷰 갱신
        await self._broadcaster.emit_session_deleted(session_id)

    async def get_catalog(self) -> dict:
        """전체 카탈로그(폴더 + 세션 배정)를 반환한다."""
        return await self._db.get_catalog()

    async def update_board_item_position(
        self,
        board_item_id: str,
        x: float,
        y: float,
    ) -> None:
        """보드 항목 좌표를 20px 격자에 스냅하여 저장한다."""
        await self._db.ensure_board_items()
        await self._db.update_board_item_position(
            board_item_id,
            self._snap_position(x),
            self._snap_position(y),
        )
        await self._broadcast_catalog()

    async def create_markdown_document(
        self,
        folder_id: str,
        title: str,
        body: str = "",
        x: Optional[float] = None,
        y: Optional[float] = None,
    ) -> dict:
        """마크다운 문서를 만들고 같은 폴더의 board item으로 배치한다."""
        document_id = str(uuid.uuid4())
        resolved_x, resolved_y = (
            (self._snap_position(x), self._snap_position(y))
            if x is not None and y is not None
            else await self._next_board_position(folder_id)
        )
        result = await self._db.create_markdown_document(
            document_id,
            folder_id,
            title,
            body,
            resolved_x,
            resolved_y,
        )
        await self._broadcast_catalog()
        return result

    async def get_markdown_document(self, document_id: str) -> Optional[dict]:
        return await self._db.get_markdown_document(document_id)

    async def update_markdown_document(
        self,
        document_id: str,
        title: Optional[str] = None,
        body: Optional[str] = None,
    ) -> Optional[dict]:
        document = await self._db.update_markdown_document(
            document_id,
            title=title,
            body=body,
        )
        await self._broadcast_catalog()
        return document

    async def delete_markdown_document(self, document_id: str) -> None:
        await self._db.delete_markdown_document(document_id)
        await self._broadcast_catalog()

    async def _validate_parent_folder(
        self,
        folder_id: str,
        parent_folder_id: Optional[str],
    ) -> None:
        if parent_folder_id is None:
            return
        if folder_id == parent_folder_id:
            raise ValueError("folder parent cycle")
        folders = await self._db.get_all_folders()
        parent_by_id = {f["id"]: f.get("parent_folder_id") for f in folders}
        current = parent_folder_id
        seen: set[str] = set()
        while current is not None:
            if current == folder_id:
                raise ValueError("folder parent cycle")
            if current in seen:
                raise ValueError("folder parent cycle")
            seen.add(current)
            current = parent_by_id.get(current)

    async def _validate_folder_parent_updates(
        self,
        parent_updates: dict[str, Optional[str]],
    ) -> None:
        if not parent_updates:
            return
        folders = await self._db.get_all_folders()
        parent_by_id = {f["id"]: f.get("parent_folder_id") for f in folders}
        parent_by_id.update(parent_updates)

        for folder_id, parent_folder_id in parent_updates.items():
            if parent_folder_id is None:
                continue
            if folder_id == parent_folder_id:
                raise ValueError("folder parent cycle")
            current = parent_folder_id
            seen: set[str] = set()
            while current is not None:
                if current == folder_id:
                    raise ValueError("folder parent cycle")
                if current in seen:
                    raise ValueError("folder parent cycle")
                seen.add(current)
                current = parent_by_id.get(current)

    # --- 폴더 시스템 프롬프트 ---

    async def get_folder_system_prompt(self, folder_id: str) -> Optional[str]:
        """폴더의 시스템 프롬프트(folderPrompt)를 반환한다.

        Args:
            folder_id: 폴더 ID.

        Returns:
            folderPrompt 문자열, 설정되지 않았으면 None.

        Raises:
            ValueError: 폴더가 존재하지 않으면.
        """
        folder = await self._db.get_folder(folder_id)
        if folder is None:
            raise ValueError(f"Folder not found: {folder_id}")
        return (folder.get("settings") or {}).get("folderPrompt")

    async def set_folder_system_prompt(
        self, folder_id: str, prompt: Optional[str]
    ) -> None:
        """폴더의 시스템 프롬프트(folderPrompt)를 설정하거나 삭제한다.

        빈 문자열 또는 None을 전달하면 folderPrompt 키를 삭제한다.
        다른 settings 키는 보존된다.

        Args:
            folder_id: 폴더 ID.
            prompt: 설정할 프롬프트. 빈 문자열 또는 None이면 삭제.

        Raises:
            ValueError: 폴더가 존재하지 않으면.
        """
        folder = await self._db.get_folder(folder_id)
        if folder is None:
            raise ValueError(f"Folder not found: {folder_id}")
        settings = dict(folder.get("settings") or {})
        if prompt:
            settings["folderPrompt"] = prompt
        else:
            settings.pop("folderPrompt", None)
        await self.update_folder(folder_id, settings=settings)

File: soul-server-ts/src/catalog/catalog_service.ts
/**
 * CatalogService — 폴더·세션 카탈로그 mutation 정본 (본 카드 신규).
 *
 * Python `packages/soul-common/src/soul_common/catalog/catalog_service.py` 키 호환 포팅.
 * MCP cogito 도구(`mcp_catalog.py`)와 dashboard 양쪽이 *같은 service*를 경유하여
 * 정책(broadcast 시점, ID 생성 책임)을 단일 자리에 둔다 (design-principles §3).
 *
 * 의존:
 *   - SessionDB — folders/sessions 테이블 mutation
 *   - SessionBroadcaster — catalog_updated / session_deleted wire emit
 *
 * 본 PR은 stored procedure 호출만 — schema DDL 정본은 `packages/db-schema/sql/schema.sql`.
 */

import { randomUUID } from "node:crypto";

import type { BoardYjsService } from "../collaboration/board_yjs_service.js";
import type {
  BoardYjsContainerRef,
  CatalogBoardItemRow,
  ListSessionSummaryRow,
  SessionDB,
} from "../db/session_db.js";
import { assertMutableFolder } from "../system_folders.js";
import type { SessionBroadcaster } from "../upstream/session_broadcaster.js";
import {
  CatalogBoardItemService,
  type CatalogBoardItemMoveResult,
} from "./catalog_board_item_service.js";

export interface CatalogFolderDto {
  id: string;
  name: string;
  sortOrder: number;
  settings: Record<string, unknown>;
  parentFolderId: string | null;
  createdAt?: string;
}

export interface BrowseFolderSessionDto {
  sessionId: string;
  title: string;
  displayName: string | null;
  status: string | null;
  sessionType: string | null;
  createdAt: string;
  updatedAt: string;
  eventCount: number;
  awaySummary: string | null;
  callerSessionId: string | null;
  nodeId: string | null;
  lastEventId: number | null;
  lastReadEventId: number | null;
}

export interface BrowseFolderResult {
  folderId: string;
  folder: CatalogFolderDto;
  childFolders: CatalogFolderDto[];
  sessions: BrowseFolderSessionDto[];
  sessionsPage: {
    cursor: number;
    limit: number;
    total: number;
    nextCursor: number | null;
  };
  boardItems: CatalogBoardItemRow[];
  counts: {
    childFolders: number;
    sessions: number;
    boardItems: number;
    documents: number;
    assets: number;
  };
}

/**
 * CatalogService — TaskManager·MCP 도구 양쪽 진입점이 공유.
 *
 * 본 클래스는 *broadcast 정책* 정본:
 *   - 폴더 CUD 후 → `broadcastCatalog()` 자동 호출 (대시보드 즉시 반영)
 *   - 세션 rename 후 → `broadcastCatalog()`
 *   - 세션 삭제 후 → `broadcastCatalog()` + `emitSessionDeleted()` (Python `delete_session`
 *     L135-141 정합)
 *
 * 도구·dashboard 진입점은 본 클래스를 호출하기만 한다 — broadcaster를 직접 호출하지 않는다.
 */
export class CatalogService {
  private readonly boardItems: CatalogBoardItemService;

  constructor(
    private readonly db: SessionDB,
    private readonly broadcaster: SessionBroadcaster,
    boardYjsService?: BoardYjsService,
  ) {
    this.boardItems = new CatalogBoardItemService(
      db,
      boardYjsService,
      () => this.broadcastCatalog(),
    );
  }

  async listFolders(): Promise<CatalogFolderDto[]> {
    const folders = await this.db.getAllFolders();
    return folders.map((f) => {
      const createdAt = f.created_at instanceof Date ? f.created_at.toISOString() : f.created_at;
      return {
        id: f.id,
        name: f.name,
        sortOrder: f.sort_order,
        settings: f.settings ?? {},
        parentFolderId: f.parent_folder_id,
        ...(createdAt ? { createdAt } : {}),
      };
    });
  }

  async listChildFolders(folderId: string | null): Promise<CatalogFolderDto[]> {
    const folders = await this.listFolders();
    return folders.filter((folder) => folder.parentFolderId === folderId);
  }

  /**
   * MCP browse용 읽기 스냅샷.
   *
   * "폴더 안에 무엇이 있나"는 folders, sessions, board_items 세 정본을 함께 봐야 한다.
   * 이 메서드는 mutation 없이 직접 자식 폴더, 세션 페이지, 문서/파일 보드 항목을 한 번에
   * 반환하여 MCP 호출자가 여러 도구를 조합하다가 누락을 만들지 않게 한다.
   */
  async browseFolder(params: {
    folderId: string;
    sessionCursor?: number;
    sessionLimit?: number;
  }): Promise<BrowseFolderResult> {
    const cursor = Math.max(0, Math.trunc(params.sessionCursor ?? 0));
    const limit = Math.min(
      100,
      Math.max(1, Math.trunc(params.sessionLimit ?? 20)),
    );
    const catalog = await this.db.getCatalog();
    const folder = catalog.folders.find((candidate) => candidate.id === params.folderId);
    if (!folder) {
      throw new Error(`folder not found: ${params.folderId}`);
    }
    const childFolders = catalog.folders.filter(
      (candidate) => candidate.parentFolderId === params.folderId,
    );
    const boardItems = catalog.boardItems.filter(
      (item) => item.folderId === params.folderId,
    );
    const { sessions, total } = await this.db.listSessionsSummary({
      search: null,
      limit,
      offset: cursor,
      folderId: params.folderId,
      nodeId: null,
    });
    return {
      folderId: params.folderId,
      folder,
      childFolders,
      sessions: sessions.map(toBrowseFolderSession),
      sessionsPage: {
        cursor,
        limit,
        total,
        nextCursor: cursor + limit < total ? cursor + limit : null,
      },
      boardItems,
      counts: {
        childFolders: childFolders.length,
        sessions: total,
        boardItems: boardItems.length,
        documents: boardItems.filter((item) => item.itemType === "markdown").length,
        assets: boardItems.filter((item) => item.itemType === "asset").length,
      },
    };
  }

  /**
   * 새 폴더 생성. id는 randomUUID로 발급 (Python `folder_create`는 외부에서 id를 받음
   * — Python catalog_service.create_folder L63-68 정합).
   */
  async createFolder(
    name: string,
    sortOrder = 0,
    parentFolderId: string | null = null,
  ): Promise<CatalogFolderDto> {
    const id = randomUUID();
    await this.assertParentAllowed(id, parentFolderId);
    await this.db.createFolder(id, name, sortOrder, parentFolderId);
    await this.broadcastCatalog();
    return { id, name, sortOrder, settings: {}, parentFolderId };
  }

  async renameFolder(folderId: string, name: string): Promise<void> {
    assertMutableFolder(folderId, "renamed");
    await this.db.updateFolder(folderId, ["name"], [name]);
    await this.broadcastCatalog();
  }

  async deleteFolder(folderId: string): Promise<void> {
    assertMutableFolder(folderId, "deleted");
    await this.db.deleteFolderById(folderId);
    await this.broadcastCatalog();
  }

  /**
   * 세션 다수를 폴더로 이동. folderId=null → 폴더 해제.
   *
   * Python `catalog_service.move_sessions_to_folder` L112-120 정합.
   */
  async moveSessionsToFolder(
    sessionIds: string[],
    folderId: string | null,
  ): Promise<void> {
    for (const sessionId of sessionIds) {
      await this.db.assignSessionToFolder(sessionId, folderId);
    }
    await this.db.ensureBoardItems();
    await this.broadcastCatalog();
  }

  /**
   * 세션 표시 이름 갱신. displayName=null → 이름 제거.
   *
   * Python `catalog_service.rename_session` L126-133 정본:
   *   db.rename_session + broadcast_catalog().
   *
   * 정규화(trim·empty→null) 책임은 호출자 (도구 핸들러 또는 dashboard API).
   */
  async renameSession(
    sessionId: string,
    displayName: string | null,
  ): Promise<void> {
    await this.db.renameSession(sessionId, displayName);
    await this.broadcastCatalog();
  }

  /**
   * 세션 삭제 — 이벤트까지 함께 삭제 (schema.sql session_delete cascade).
   *
   * Python `catalog_service.delete_session` L135-141 정합:
   *   db.delete_session + broadcast_catalog() + emit_session_deleted.
   *
   * 두 wire를 모두 발사하는 이유:
   *   - catalog_updated → 폴더 트리 갱신
   *   - session_deleted → 세션 목록 행 즉시 제거
   */
  async deleteSession(sessionId: string): Promise<void> {
    await this.db.deleteSession(sessionId);
    await this.broadcastCatalog();
    await this.broadcaster.emitSessionDeleted(sessionId);
  }

  /**
   * 폴더 system prompt 조회 — settings.folderPrompt를 읽는다.
   *
   * Python `catalog_service.get_folder_system_prompt` 정합. 폴더 부재 시 throw.
   */
  async getFolderSystemPrompt(folderId: string): Promise<string | null> {
    const folder = await this.db.getFolderById(folderId);
    if (!folder) {
      throw new Error(`folder not found: ${folderId}`);
    }
    const settings = folder.settings;
    const prompt = settings.folderPrompt;
    return typeof prompt === "string" ? prompt : null;
  }

  /**
   * 폴더 system prompt 설정. prompt=null 또는 빈 문자열 → settings.folderPrompt 제거.
   *
   * Python `catalog_service.set_folder_system_prompt` 정합. folder_update stored proc로
   * settings 컬럼 전체 갱신 (기존 키 유지 + folderPrompt만 갱신).
   */
  async setFolderSystemPrompt(
    folderId: string,
    prompt: string | null,
  ): Promise<void> {
    const folder = await this.db.getFolderById(folderId);
    if (!folder) {
      throw new Error(`folder not found: ${folderId}`);
    }
    const settings: Record<string, unknown> = { ...folder.settings };
    if (prompt && prompt.trim().length > 0) {
      settings.folderPrompt = prompt;
    } else {
      delete settings.folderPrompt;
    }
    await this.db.updateFolder(
      folderId,
      ["settings"],
      [JSON.stringify(settings)],
    );
    await this.broadcastCatalog();
  }

  async setFolderParent(
    folderId: string,
    parentFolderId: string | null,
  ): Promise<void> {
    assertMutableFolder(folderId, "moved");
    await this.assertParentAllowed(folderId, parentFolderId);
    await this.db.updateFolder(
      folderId,
      ["parent_folder_id"],
      [parentFolderId],
    );
    await this.broadcastCatalog();
  }

  /**
   * 카탈로그 wire 발사 — Python `catalog_service._broadcast_catalog` L39-47 정합.
   * 폴더 트리·세션 매핑을 한 번에 broadcast하여 대시보드가 일관된 스냅샷으로 갱신.
   */
  async broadcastCatalog(): Promise<void> {
    const catalog = await this.db.getCatalog();
    await this.broadcaster.emitCatalogUpdated(catalog);
  }

  async updateBoardItemPosition(
    boardItemId: string,
    x: number,
    y: number,
  ): Promise<void> {
    await this.boardItems.updateBoardItemPosition(boardItemId, x, y);
  }

  async moveBoardItemToContainer(params: {
    boardItemId: string;
    target: BoardYjsContainerRef;
    position?: { x: number; y: number };
    idempotencyKey: string;
  }): Promise<CatalogBoardItemMoveResult> {
    return await this.boardItems.moveBoardItemToContainer(params);
  }

  async createMarkdownDocument(params: {
    folderId: string;
    container?: BoardYjsContainerRef | null;
    title: string;
    body?: string;
    x?: number;
    y?: number;
  }): Promise<Awaited<ReturnType<SessionDB["createMarkdownDocument"]>>> {
    return await this.boardItems.createMarkdownDocument(params);
  }

  async getMarkdownDocument(documentId: string) {
    return await this.boardItems.getMarkdownDocument(documentId);
  }

  async updateMarkdownDocument(
    documentId: string,
    fields: { title?: string; body?: string; expectedVersion: number },
  ) {
    return await this.boardItems.updateMarkdownDocument(documentId, fields);
  }

  async deleteMarkdownDocument(documentId: string): Promise<void> {
    await this.boardItems.deleteMarkdownDocument(documentId);
  }

  private async assertParentAllowed(
    folderId: string,
    parentFolderId: string | null,
  ): Promise<void> {
    if (parentFolderId === null) return;
    if (folderId === parentFolderId) {
      throw new Error("folder parent cycle");
    }
    const folders = await this.db.getAllFolders();
    const parentById = new Map(folders.map((folder) => [folder.id, folder.parent_folder_id]));
    let current: string | null | undefined = parentFolderId;
    const seen = new Set<string>();
    while (current) {
      if (current === folderId) {
        throw new Error("folder parent cycle");
      }
      if (seen.has(current)) {
        throw new Error("folder parent cycle");
      }
      seen.add(current);
      current = parentById.get(current);
    }
  }
}

function toBrowseFolderSession(row: ListSessionSummaryRow): BrowseFolderSessionDto {
  return {
    sessionId: row.session_id,
    title: row.display_name ?? row.session_id,
    displayName: row.display_name,
    status: row.status,
    sessionType: row.session_type,
    createdAt: serializeDate(row.created_at),
    updatedAt: serializeDate(row.updated_at),
    eventCount: row.event_count,
    awaySummary: row.away_summary,
    callerSessionId: row.caller_session_id,
    nodeId: row.node_id,
    lastEventId: row.last_event_id,
    lastReadEventId: row.last_read_event_id,
  };
}

function serializeDate(value: Date | string): string {
  if (value instanceof Date) return value.toISOString();
  const ms = Date.parse(value);
  return Number.isFinite(ms) ? new Date(ms).toISOString() : value;
}
